### ISSUE TITLE
`evil-define-key' for minor mode does not take effect until a state transition

### ISSUE THREAD

**Issue by TheBB** [emacs-evil/evil#301]
Originally reported by: **Anonymous**

----------------------------------------

Unit test. (you can change flyspell to any toggleble minor mode)

(require 'flyspell)
(evil-define-key 'normal flyspell-mode-map [f5] 
   (lambda () (interactive) (message "Hey")))

In *scratch* buffer

M-x: flyspell-mode

press F5, key is not bound. 
press i then esc, key is now bound

This is a problem because evil is enabled from find-file-hook. before major mode specific hooks are called. If major mode specific hooks enable any minor modes, evil is already initialized, and it does not see those minor modes until a state transition is made, so any bindings for them are in-active

Tested with

GNU Emacs 24.3.1 (x86_64-suse-linux-gnu, X toolkit, Xaw scroll bars) of 2013-05-26 on momoland

Evil from git 8f07f5b0b1fc888428b913e14944e89a0341e7b5

----------------------------------------
- Bitbucket: https://bitbucket.org/lyro/evil/issue/301

**Comment by fiddlerwoaroof** [comment]
Is there any progress/workarounds on this bug?  It just started being an issue for me after an upgrade

**Comment by noctuid** [comment]
Use `evil-define-minor-mode-key` or call `evil-normalize-keymaps` in the relevant mode hook.

**Comment by fiddlerwoaroof** [comment]
I solved the issue by calling `evil-normalize-keymaps` In the `window-configuration-change-hook`

**Comment by mohkale** [comment]
I think I'm encountering this with tempel but hooking evil-normalize-keymaps after the relevent functions that update the minor mode map doesn't fix the issue. Any tips on why?

So far I've tried.

```elisp
(advice-add 'tempel--insert :after #'evil-normalize-keymaps)
(advice-add 'tempel--done :after #'evil-normalize-keymaps)
```

**Comment by tomdl89** [comment]
@mohkale when you say you think you are encountering this with tempel, what makes you think that? Does hitting the esc key fix the issue?

**Comment by mohkale** [comment]
Yes. I hit tempel expand, none of the bindings in tempel map are active. I hit escape and back to insert and they suddenly work again.

**Comment by mohkale** [comment]
Because tempel is a snippet library not a mode and tempel-map is only active when a snippet is expanded and being edited. Sorry if that wasn't clear in my original description.

**Comment by tomdl89** [comment]
Ah sorry - no I'm not familiar with tempel. Can you share the code you're using to define the keybindings then? I will try to repro the problem and find a solution from that.

**Comment by mohkale** [comment]
BTW I fixed the issue with tempel by doing:

```elisp
(use-package tempel
  :preface
  (defun tempel--normalize-evil-keymaps+ (&rest _)
    (evil-normalize-keymaps))
  :init
  (advice-add 'tempel--insert :after #'tempel--normalize-evil-keymaps+)
  (advice-add 'tempel-done :after #'tempel--normalize-evil-keymaps+))
```
